mypyc/genops.py

```diff
@@ -1163,23 +1362,31 @@ def for_loop_helper(self, index: Lvalue, expr: Expression,
         if (isinstance(expr, CallExpr)
                 and isinstance(expr.callee, RefExpr)
                 and expr.callee.fullname == 'builtins.range'):
+            # TODO: yield statements do not work inside range-based loops.
+
             condition_block = BasicBlock()
             self.push_loop_stack(increment_block, exit_block)
 
             # Special case for x in range(...)
             # TODO: Check argument counts and kinds; check the lvalue
             end = expr.args[0]
-            end_reg = self.accept(end)
 
             # Initialize loop index to 0.
-            index_target = self.get_assignment_target(index)
-            self.assign(index_target, self.accept(IntExpr(0)), line)
+            if self.fn_info.is_generator:
+                assert isinstance(index, NameExpr)
+                assert index.node is not None
+                index_target = self.add_var_to_env_class(index.node, int_rprimitive,
+                                                         self.fn_info.generator_class)
+                self.assign(index_target, self.add(LoadInt(0)), line)
+            else:
+                index_target = self.get_assignment_target(index)
+                self.assign(index_target, self.accept(IntExpr(0)), line)
```

Comment:
Why is any of this necessary? The index is already an AssignmentTarget.

Reply:
Fixed.